### PR TITLE
Update auto optimizer strategy and fix mismatched `warmup_lr`

### PR DESCRIPTION
**Pull request by Laughing-q**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Improves optimizer setup and warmup logic by labeling parameter groups explicitly and refining MuSGD defaults for more reliable training ⚙️🚀

### 📊 Key Changes
- **Added explicit `param_group` labels** to optimizer parameter groups (`"bias"`, `"weight"`, `"bn"`, `"muon"`) to make group-specific behavior deterministic 🏷️
- **Warmup learning-rate logic now targets bias params by label** (`param_group == "bias"`) instead of assuming the first param group is bias (index-based) 🔥
- **Auto-optimizer selection simplified to always use MuSGD** (removes switching between `"SGD"` and `"MuSGD"` based on iteration count) 🤖
- **Introduced dynamic MuSGD mixing factors (`muon`, `sgd`) based on training length**, and passed them via `functools.partial` when constructing the optimizer 🎛️
- **Clarified MuSGD fine-tuning behavior** in comments for the “higher LR for certain parameters” logic 📝

### 🎯 Purpose & Impact
- **More robust training warmup**: avoids subtle bugs where bias warmup could be misapplied if param group ordering changes (common when optimizers or grouping logic evolves) ✅
- **More predictable optimizer behavior**: explicit group labeling makes training configs easier to reason about and reduces “works by coincidence” assumptions 🧩
- **Potential training-quality and stability improvements with MuSGD**: dynamic `muon/sgd` scaling adapts behavior for longer vs shorter runs, which may improve convergence and fine-tuning experience 📈
- **Minor behavior change**: runs that previously auto-selected plain SGD may now use MuSGD instead, which could slightly change learning dynamics and results (even with similar hyperparameters) ⚠️

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Laughing-q, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message, and an Ultralytics engineer will assist with review soon. 🧩

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -408,10 +409,15 @@ def _do_train(self):
...
                     self.accumulate = max(1, int(np.interp(ni, xi, [1, self.args.nbs / self.batch_size]).round()))
-                    for j, x in enumerate(self.optimizer.param_groups):
+                    for x in self.optimizer.param_groups:
                         # Bias lr falls from 0.1 to lr0, all other lrs rise from 0.0 to lr0
                         x["lr"] = np.interp(
-                            ni, xi, [self.args.warmup_bias_lr if j == 0 else 0.0, x["initial_lr"] * self.lf(epoch)]
+                            ni,
+                            xi,
+                            [
+                                self.args.warmup_bias_lr if x["param_group"] == "bias" else 0.0,
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py]: ⚠️ **HIGH**: `x["param_group"]` will raise a `KeyError` if an optimizer param group is missing this tag (e.g., custom optimizer, older checkpoints, or externally-constructed optimizer). Warmup should be resilient and fall back gracefully (e.g., treat missing tag as non-bias).

**Suggested change:**
```suggestion
                                self.args.warmup_bias_lr if x.get("param_group") == "bias" else 0.0,
```

File: ultralytics/engine/trainer.py [hunk 2]
```diff
@@ -991,7 +998,7 @@ def build_optimizer(self, model, name="auto", lr=0.001, momentum=0.9, decay=1e-5
                 p2 = [v for k, v in p.items() if not pattern.search(k)]
                 g_.extend([{"params": p1, **x, "lr": lr * 3}, {"params": p2, **x}])
             g = g_
-        optimizer = getattr(optim, name, MuSGD)(params=g)
+        optimizer = getattr(optim, name, partial(MuSGD, muon=muon, sgd=sgd))(params=g)
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:1001]: 💡 **MEDIUM**: Using `getattr(optim, name, partial(MuSGD, ...))` means any unknown/typo'd optimizer name will silently fall back to `MuSGD` (now with iteration-dependent `muon/sgd` scaling), which can make misconfiguration hard to detect. Consider only applying the `partial(MuSGD, ...)` path when `name == "MuSGD"` (or `use_muon`) and otherwise error out if the optimizer isn't found.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good direction overall: tagging param groups makes warmup LR behavior deterministic and avoids relying on param-group ordering. Two issues to address: (1) warmup currently assumes every param group has `param_group` and can crash with a `KeyError`; using `.get()` avoids brittle behavior. (2) the optimizer factory still silently falls back to MuSGD for unknown names, and with the new `partial()` scaling this becomes even harder to spot—consider tightening this to fail fast on invalid optimizer names.

💬 Posted 2 inline comments

**Review by lmycross — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
Merged—awesome work @Laughing-q (and thanks to @lmycross for the contributions)! This is a really solid quality upgrade to the training stack: explicit `param_group` labels, bias warmup keyed by label instead of ordering, and a clearer, more deterministic MuSGD path all help make optimizer behavior more predictable and resilient as the code evolves.

As **Leonardo da Vinci** put it: *“Simplicity is the ultimate sophistication.”* This PR embodies that—by removing “works by coincidence” assumptions and making intent explicit, you’ve improved reliability and made future iteration on YOLO11 training behavior safer and easier to reason about.

Appreciate the thoughtful engineering and careful attention to training stability—thanks for moving Ultralytics forward.
